syft/create_sbom_config.go

```diff
@@ -0,0 +1,312 @@
+// selected and the tokens that were sensitive to this selection (both for adding and removing from the final set).
+func (c CreateSBOMConfig) makeTaskGroups(src source.Description) ([][]task.Task, *catalogerManifest, error) {
+	var taskGroups [][]task.Task
+
+	// generate package and file tasks based on the configuration
+	environmentTasks := c.environmentTasks()
+	relationshipsTasks := c.relationshipTasks(src)
+	fileTasks := c.fileTasks()
+	pkgTasks, selectionEvidence, err := c.packageTasks(src)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// combine the user-provided and configured tasks
+	if c.Files.Selection == file.OwnedFilesSelection {
+		// special case: we need the package info when we are cataloging files owned by packages
+		taskGroups = append(taskGroups, pkgTasks, fileTasks)
+	} else {
+		taskGroups = append(taskGroups, append(pkgTasks, fileTasks...))
+	}
+
+	// all relationship work must be done after all nodes (files and packages) have been cataloged
+	if len(relationshipsTasks) > 0 {
+		taskGroups = append(taskGroups, relationshipsTasks)
+	}
+
+	// identifying the environment (i.e. the linux release) must be done first as this is required for package cataloging
+	taskGroups = append(
+		[][]task.Task{
+			environmentTasks,
+		},
+		taskGroups...,
+	)
+
+	return taskGroups, &catalogerManifest{
+		Requested: selectionEvidence.Request,
+		Used:      formatTaskNames(pkgTasks),
+	}, nil
+}
+
+// fileTasks returns the set of tasks that should be run to catalog files.
+func (c CreateSBOMConfig) fileTasks() []task.Task {
+	var tsks []task.Task
+
+	if t := task.NewFileDigestCatalogerTask(c.Files.Selection, c.Files.Hashers...); t != nil {
+		tsks = append(tsks, t)
+	}
+	if t := task.NewFileMetadataCatalogerTask(c.Files.Selection); t != nil {
+		tsks = append(tsks, t)
+	}
+	return tsks
+}
+
+// packageTasks returns the set of tasks that should be run to catalog packages.
+func (c CreateSBOMConfig) packageTasks(src source.Description) ([]task.Task, *task.Selection, error) {
+	cfg := task.CatalogingFactoryConfig{
+		SearchConfig:         c.Search,
+		RelationshipsConfig:  c.Relationships,
+		DataGenerationConfig: c.DataGeneration,
+		PackagesConfig:       c.Packages,
+	}
+	tsks, err := c.packageTaskFactories.Tasks(cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create package cataloger tasks: %w", err)
+	}
+
+	if len(c.CatalogerSelection.DefaultNamesOrTags) == 0 {
+		defaultTag, err := findDefaultTag(src)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to determine default cataloger tag: %w", err)
+		}
+
+		if defaultTag != "" {
+			c.CatalogerSelection.DefaultNamesOrTags = append(c.CatalogerSelection.DefaultNamesOrTags, defaultTag)
+		}
+
+		c.CatalogerSelection.RemoveNamesOrTags = replaceDefaultTagReferences(defaultTag, c.CatalogerSelection.RemoveNamesOrTags)
+		c.CatalogerSelection.SubSelectTags = replaceDefaultTagReferences(defaultTag, c.CatalogerSelection.SubSelectTags)
+	}
```

Comment:
comment from @willmurphyscode this should probably be decomposed into a separate function (since it's a lower level of abstraction)